File: 2265-count-nodes-equal-to-average-of-subtree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        matching_count = 0

        def dfs(node: Optional[TreeNode]) -> tuple[int, int]:
            nonlocal matching_count
            if not node:
                return 0, 0

            # Post-order traversal: collect data from left and right children
            left_sum, left_cnt = dfs(node.left)
            right_sum, right_cnt = dfs(node.right)

            # Calculate total sum and node count for current subtree
            total_sum = left_sum + right_sum + node.val
            total_cnt = left_cnt + right_cnt + 1

            # Check if node's value equals average of subtree (rounded down)
            if node.val == total_sum // total_cnt:
                matching_count += 1

            return total_sum, total_cnt

        dfs(root)
        return matching_count

File: 2265-count-nodes-equal-to-average-of-subtree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int matchingNodesCount = 0;

    // Helper function returns pair<subtree_sum, subtree_count>
    pair<int, int> dfs(TreeNode* node) {
        if (!node) {
            return {0, 0};
        }

        // Compute subproblems for left and right subtrees
        auto [leftSum, leftCount] = dfs(node->left);
        auto [rightSum, rightCount] = dfs(node->right);

        // Aggregate results for current node
        int currentSum = leftSum + rightSum + node->val;
        int currentCount = leftCount + rightCount + 1;

        // Check condition (integer division automatically rounds down for non-negative integers)
        if (node->val == currentSum / currentCount) {
            matchingNodesCount++;
        }

        return {currentSum, currentCount};
    }

public:
    int averageOfSubtree(TreeNode* root) {
        matchingNodesCount = 0;
        dfs(root);
        return matchingNodesCount;
    }
};